src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -325,14 +308,193 @@ async def debug_error_string(self) -> Optional[str]:
+    _intercepted_response_iterator: Optional[AsyncIterable[ResponseType]]
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, interceptors: Sequence[UnaryStreamClientInterceptor],
+                 request: RequestType, timeout: Optional[float],
+                 metadata: MetadataType,
+                 credentials: Optional[grpc.CallCredentials],
+                 wait_for_ready: Optional[bool], channel: cygrpc.AioChannel,
+                 method: bytes, request_serializer: SerializingFunction,
+                 response_deserializer: DeserializingFunction,
+                 loop: asyncio.AbstractEventLoop) -> None:
+        self._loop = loop
+        self._channel = channel
+        self._response_aiter = self._wait_for_interceptor_task_response_iterator()
+        self._call_response_iterator = CallResponseIterator()
+        self._intercepted_response_iterator = None
+        interceptors_task = loop.create_task(
+            self._invoke(interceptors, method, timeout, metadata,
+                         credentials, wait_for_ready, request,
+                         request_serializer, response_deserializer)
+        )
+        super().__init__(interceptors_task)
+
+    # pylint: disable=too-many-arguments
+    async def _invoke(self,
+                      interceptors: Sequence[UnaryUnaryClientInterceptor],
+                      method: bytes, timeout: Optional[float],
+                      metadata: Optional[MetadataType],
+                      credentials: Optional[grpc.CallCredentials],
+                      wait_for_ready: Optional[bool], request: RequestType,
+                      request_serializer: SerializingFunction,
+                      response_deserializer: DeserializingFunction
+                     ) -> UnaryStreamCall:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryStreamClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType,
+                response_iterator: AsyncIterable[ResponseType]
+               ) -> _base_call.UnaryUnaryCall:
+
+            interceptor = next(interceptors, None)
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+
+                return await interceptor.intercept_unary_stream(
+                    continuation, client_call_details, request, response_iterator)
+            else:
+                call = UnaryStreamCall(
+                    request, _timeout_to_deadline(client_call_details.timeout),
+                    client_call_details.metadata,
+                    client_call_details.credentials,
+                    client_call_details.wait_for_ready, self._channel,
+                    client_call_details.method, request_serializer,
+                    response_deserializer, self._loop)
+
+                self._call_response_iterator.set_call(call)
```

Comment:
Is `_call_response_iterator` necessary? Can we use the `call` object directly as response iterator?

Reply:
Agree. I'm not clear on the purpose of the `CallResponseIterator` in general.

Reply:
Is about how the interface is being designed, we provide the following signature:

```python
    async def intercept_unary_stream(
            self, continuation: Callable[[ClientCallDetails, RequestType, AsyncIterable[ResponseType]],
```

So any interceptor can interpose its own iterator by using the iterator that is provided as a parameter. The `CallResponseIterator` allows us to provide this initial iterator, which would be finally configured once we have the call, by making the `set_call`.

We could follow another pattern by asking the user for wrapping the call that has been returned and returning a wrapped call. For doing the second option we would need to provide a new interface for allowing this wrapping, because the object returned would need to publish the same contract as a generic call.

Also, consider that once we start implementing the Streaming Input arities we could end up by following the same pattern, so providing a new parameter called `request_iterator`. So presenting both interfaces, `request_iterator` and `response_iterator`, in the same way, would be a less cognitive burden for the user.

Reply:
Ive found out a use case where the current implementation it's not enough for the user. With the current implementation, the user is not able to access to the call since the call is not yet available until the continuation has been called.

The user would be forced to make the proper wiring for not losing the call context, for example:

```python
class ResponseIterator:
    def __init__(self, response_iterator):
        self._response_iterator = response_iterator

    def set_call(self., call):
       self.call = call

    async def my_iter(self):
        async for response in self._response_iterator:
            yield response

        print(await call.status_code())

    def __aiter__(self):
        return  my_iter()

sync def intercept_unary_stream(..., response_iterator):
    my_response_itearator = ResponseIterator(response_iterator)
    call = await continuation(..., my_response_itearator)
    my_response_iterator.set_call(call)
    return call
```

In the example, the user needs to implement their own wiring for having visibility of the status of the call once the iterator is finished.

That makes me think that we would need to provide this kind of glue, so a way that the user can wrapper easily the iterator without losing call context.

I will give it a try, by providing our own wrapper.

BTW:  we do have a problem with the `add_done_callback` sooner or later we would need to accept coroutines as callbacks, otherwise, retrieve call attributes within the context of a NORMAL FUNCION is quite complicated

Reply:
changed, PTAL.